monoio-http/src/common/request.rs

```diff
@@ -1,24 +1,90 @@
+use std::{collections::HashMap, io::Cursor};
+
+use bytes::Bytes;
+use cookie::Cookie; // Import the Cookie type from the cookie crate
+use cookie::CookieJar;
 pub use http::request::{Builder as RequestBuilder, Parts as RequestHead};
+use multipart::server::Multipart;
+use serde_urlencoded::from_str;
 
-use super::BorrowHeaderMap;
+use super::{
+    body::{Body, FixedBody},
+    error::{ExtractError, HttpError},
+    BorrowHeaderMap,
+};
 use crate::{
     common::{FromParts, IntoParts},
     h1::payload::Payload,
+    impl_cookie_extractor,
 };
 
-pub type Request<P = Payload> = http::request::Request<P>;
+#[derive(Clone)]
+pub struct Request<P = Payload> {
```

Comment:
It is possible to make this feature zero cost(no cost if not use)? Not every one need to extract cookie or query parameters.
An idea is to keep current Request and rename the new Request something like RequestWithMeta(I'm not good at naming)? And the later struct can convert to or from the former Request.

Reply:
cookie_jar and url_params will be populated only if cookie_parse or url_parse is called, should be none otherwise. 

 As the request passes through the handler chain passes there'll be a lot of conversions between Request and RequestWithMeta no?

Reply:
fn handle1 (req: Request) {
      let reqWithMeta = request.cookie_parse() // Parse cookies
      // add some more cookies
      self.handle2(reqWithMeta.into()) // Have to call serialize_cookies_into_header internally to avoid losing changes
 }
 
fn handle2(req: Request) {
      let reqWithMeta = request.cookie_parse() // Parse cookies AGAIN
      // add some more cookies
    self.handle3(reqWithMeta.into()) // Changes encoded again to header
}
------------------------------------------------------------------------

fn handle1 (req: Request) {
      let req = request.cookie_parse() // Parse cookies ONCE
      // add some more cookies
      self.handle2(req) // No serialization here
 }
 
fn handle2(req: Request) {
      // add some more cookies
    req.serialize_cookies_into_header() // User has to explicitly call fn write changes to header. Encoding ONCE
    self.handle3(req)
}

Reply:
Ok I get what you mean. If user intended to use `RequestWithMeta`, he should pass the struct in his own logic instead of convert to `Request`. Only when interact with other components(for example http client) should convert.
We can let users check if it is convinent to use `RequestWithMeta` way, if not, then we do modification at `Request` like now.

Reply:
I am proposing we only use ```monoio_http::Request```, and not introduce another struct ```RequestWithMeta```.

 ```Request``` will not parse anything on its own. It'll be Request {inner, none, none}. The user will call ```req.parse_cookies_from_header()``` whenever required, the metadata is then populated Request {inner,Some(jar), none}. You can keep passing this Request along the handler chain, any get and set operations will be only in the metadata. The headers remain unchanged. When it's time to move to a different component like monoio-http-client, like you said, we need the headers to be updated with latest values. We make the user call `req.serialize_cookies_into_header()`. 
 
 `fn serialize_cookies_into_header(&mut self) -> Result<(), HttpError> `
`fn parse_cookies_from_header(&mut self) -> Result<(), HttpError> `